flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java

```diff
@@ -570,284 +583,278 @@ private void closeCurrentPartFile(BucketState<T> bucketState) throws Exception {
-								Path pendingPath = new Path(finalPath.getParent(),
-									pendingPrefix + finalPath.getName()).suffix(pendingSuffix);
+								Path pendingPath = getPendingPathFor(finalPath);
 
 								fs.rename(pendingPath, finalPath);
 								LOG.debug(
 									"Moving pending file {} to final location having completed checkpoint {}.",
 									pendingPath,
 									pastCheckpointId);
 							}
-							checkpointsToRemove.add(pastCheckpointId);
+							pendingIt.remove();
 						}
 					}
-					if (!bucketState.isWriterOpen && bucketState.pendingFiles.isEmpty()) {
+
+					if (!bucketState.isWriterOpen &&
+						bucketState.pendingFiles.isEmpty() &&
+						bucketState.pendingFilesPerCheckpoint.isEmpty()) {
+
 						// We've dealt with all the pending files and the writer for this bucket is not currently open.
 						// Therefore this bucket is currently inactive and we can remove it from our state.
-						it.remove();
-					} else {
-						for (Long toRemove : checkpointsToRemove) {
-							bucketState.pendingFilesPerCheckpoint.remove(toRemove);
-						}
+						stateIt.remove();
 					}
 				}
 			}
 		}
 	}
 
 	@Override
-	public State<T> snapshotState(long checkpointId, long checkpointTimestamp) throws Exception {
+	public void snapshotState(FunctionSnapshotContext context) throws Exception {
+		Preconditions.checkNotNull(restoredBucketStates, "The operator has not been properly initialized.");
+
+		restoredBucketStates.clear();
+
 		synchronized (state.bucketStates) {
-			for (BucketState<T> bucketState : state.bucketStates.values()) {
+			int subtaskIdx = getRuntimeContext().getIndexOfThisSubtask();
+
+			for (Map.Entry<String, BucketState<T>> bucketStateEntry : state.bucketStates.entrySet()) {
+				BucketState<T> bucketState = bucketStateEntry.getValue();
+
 				if (bucketState.isWriterOpen) {
-					long pos = bucketState.writer.flush();
-					bucketState.currentFileValidLength = pos;
+					bucketState.currentFileValidLength = bucketState.writer.flush();
 				}
+
 				synchronized (bucketState.pendingFilesPerCheckpoint) {
-					bucketState.pendingFilesPerCheckpoint.put(checkpointId, bucketState.pendingFiles);
+					bucketState.pendingFilesPerCheckpoint.put(context.getCheckpointId(), bucketState.pendingFiles);
 				}
 				bucketState.pendingFiles = new ArrayList<>();
 			}
+			restoredBucketStates.add(state);
+
+			if (LOG.isDebugEnabled()) {
+				LOG.debug("BucketingSink idx {} checkpointed {}.", subtaskIdx, state);
+			}
 		}
-		return state;
 	}
 
-	@Override
-	public void restoreState(State<T> state) {
-		this.state = state;
-
-		try {
-			initFileSystem();
-		} catch (IOException e) {
-			LOG.error("Error while creating FileSystem in checkpoint restore.", e);
-			throw new RuntimeException("Error while creating FileSystem in checkpoint restore.", e);
-		}
+	private void handleRestoredBucketState(BucketState<T> bucketState) {
+		// we can clean all the pending files since they were renamed to
```

Comment:
I do not get the question? You mean when the thing on the comment can occur? The comment just says that we remove it from our current state. 

Now if you mean when two tasks are responsible for the same files, then consider the  following: 

Files written by a task contain the index of the task in their filename and the previous restore() was cleaning the files based on this index. When refactoring the BucketingSink we discussed with @aljoscha and we decided to not make the assumption that upon restoring/rescaling new task X will take the state of the previous task X, but states may be re-shuffled arbitrarily. This was also made in order to be future-proof.

Under this new assumption, a new task 0 may take the state of the old task 2. If the new task 2 deletes all the pending files that have its index just because they are no longer in its new state, then task 0 will not be able to recover them.